### PR TITLE
Hearthstone Spice debut

### PR DESCRIPTION
**Pull request by Akryum**
# Spice Pull Request

**What does your Instant Answer do?**
It shows infos about any Hearthstone card with a thumbnail.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It allows to get instantly main informations like what a card does, without searching in a database or a wiki.

**What is the data source for your Instant Answer? (Provide a link if possible)**
Gamepedia Hearthstone Wiki
http://hearthstone.gamepedia.com/

**Why did you choose this data source?**
It seems to be the most accurate, and has image for every card.

**Are there any other alternative (better) data sources?**
Not better alternatives:
http://hearthstonejson.com/
http://hearthstoneapi.herokuapp.com/

**What are some example queries that trigger this Instant Answer?**
hearthstone leeroy
hearthstone tidecaller
hearthstone murloc warleader

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Gamers!

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No.

**Which existing Instant Answers will this one supersede/overlap with?**
None I can think of.

**Are you having any problems? Do you need our help with anything?**
No thanks. :)

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
The API is a Mediawiki API. A Warranty disclaimer applies.

**Where did you hear about DuckDuckHack? (For first time contributors)**
On the duckduckgo timeline (https://duckduckgo.com/about). :)

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**

duckduckgo-hearthstone3

## Checklist
Please place an 'X' where appropriate.

```
[x] Added metadata and attribution information
[x] Wrote test file and added to t/ directory
[x] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[x] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[x] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [x] Google Chrome
        [x] Firefox
        [x] Opera
        [] IE 8
        [] IE 9
        [] IE 10
(Tested on IE11)

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [x] Firefox
        [] Native Browser
        [x] Google Chrome
        [] Opera
```

### CODE REVIEW

File: share/spice/hearthstone/hearthstone.handlebars [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        <p>
+            <span>Cost: <b>{{cost}}</b> <image class="zci--hearthstone--icon" src="http://hydra-media.cursecdn.com/hearthstone.gamepedia.com/8/8b/Mana_icon.png"/></span>
+            
+            {{#if attack}}
+            <span> - Attack: <b>{{attack}}</b> <image class="zci--hearthstone--icon" src="http://hydra-media.cursecdn.com/hearthstone.gamepedia.com/f/fe/Attack_icon.png"/></span>
+            {{/if}}
+            
+            {{#if health}}
+            <span> - Health: <b>{{health}}</b> <image class="zci--hearthstone--icon" src="http://hydra-media.cursecdn.com/hearthstone.gamepedia.com/1/17/Health_icon.png"/></span>
+            {{/if}}
```
Comment (MrChrisW) [inline share/spice/hearthstone/hearthstone.handlebars]: You'll need to use the `imageProxy` helper for all images, see https://duck.co/duckduckhack/spice_handlebars_helpers#imageproxy

File: share/spice/hearthstone/hearthstone.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        // Card Hero
+        if(card.hero === "Any") {
+            card.hero = "Any Hero";
+        }
+        
+        
+        // Render the response
+        Spice.add({
+            id: "hearthstone",
+            name: "Card",
```
Comment (MrChrisW) [inline share/spice/hearthstone/hearthstone.js]: Thinking this should be update this to "Hearthstone", "Hearthstone Card" or the standard "Answer".

File: .travis.yml [hunk 3]
```diff
@@ -1,5 +1,4 @@
 ---
-sudo: false
 before_install:
   - export HARNESS_OPTIONS=j10:c HARNESS_TIMER=1
   - rm .travis.yml
```
Comment (MrChrisW) [inline .travis.yml]: This file was changed recently in the master branch. Looks like your master branch hasn't been updated with those changes. Probably the easiest option is to manually add the `sudo: false` line to your local file and commit. Another option is to perform a rebase.

File: share/spice/hearthstone/hearthstone.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            name: api_result.Has_name,
+            type: api_result.Has_card_type,
+            rarity: api_result.Has_rarity,
+            hero: api_result.Has_class,
+            cost: api_result.Has_mana_cost,
+            attack: api_result.Has_attack,
+            health: api_result.Has_health,
+            description: api_result.Has_description,
+            flavor: api_result.Has_flavor_text
+        };
```
Comment (MrChrisW) [inline share/spice/hearthstone/hearthstone.js]: Its it possible the API won't return one of any of the values above?

**Comment by MrChrisW** [comment]
@Akryum This is a great first pass! Thanks for contributing :cake:  I've added some feedback and I'm sure other community leaders will check this out shortly. Feel free to ask any questions. Do you own the API endpoint? bytevortex.net

**Comment by Akryum** [comment]
@MrChrisW Thanks for the feedback! And yes I own bytevortex.net ;)

**Comment by MrChrisW** [comment]
@Akryum No problem! Happy to help. Great!

**Comment by Akryum** [comment]
Feedback applied! :)

**Comment by MrChrisW** [comment]
Thanks @Akryum, We'll just need to await a staff review or other feedback from the community!

**Comment by Akryum** [comment]
Ironed out a few quirks.

**Comment by MrChrisW** [comment]
@Akryum :+1:

**Comment by Akryum** [comment]
Got a question: How are IA supposed to be translated/regionalized?

**Comment by MrChrisW** [comment]
@Akryum Good question! 

Right now all Instant Answers provide English results. Translation will come later, so initially we're moving forward with just English answers.

**Comment by moollaza** [comment]
@Akryum @MrChrisW 

This is looking pretty good so far, I have 2 concerns:

1. We're using the base template and a lot of custom HTML. Let's use a pre-defined template group (I think the `info` group should work here) and reduce the amount of custom code for this. I think  the look/feel of the new [MTG Spice](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/mtg/mtg.js) https://duckduckgo.com/?q=mtg+nullify&ia=magicthegathering would be ideal. It also makes use of the InfoBox so we have room to show lots of data with minimal effort and templating.

2. We're not doing any relevancy checking against the query. Right now this will trigger whenever the query has "hearthstone" **anywhere** in the query. I'm sure a large portion of those potential queries aren't a search for a hearthstone card (e.g "hearthstone game", "hearthstone instructions", "hearthstone stoves")

    Let's improve our triggers and add some checking so we're confident the query that triggered this is likely to be a search for a hearthstone card and then we don't have to worry about showing lots of potentially irrelevant results.

File: share/spice/hearthstone/hearthstone.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        // Validate the response
+        if (!api_result || api_result.error || !api_result.Has_name || !api_result.page) {
+            return Spice.failed('hearthstone');
+        }
+        
+        // Card Data
+        // May (but very unlikely) return <undefined> values for : image_url, Has_card_type, Has_rarity,
+        // Has_class, Has_mana_cost
+        var card = {
```
Comment (moollaza) [inline share/spice/hearthstone/hearthstone.js]: This should be handled with a `normalize` function: https://duck.co/duckduckhack/spice_displaying#data-normalization

File: share/spice/hearthstone/hearthstone.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            page: api_result.page,
+            image: api_result.image_url,
+            name: api_result.Has_name,
+            type: api_result.Has_card_type,
+            rarity: api_result.Has_rarity,
+            hero: api_result.Has_class,
+            cost: api_result.Has_mana_cost,
+            attack: api_result.Has_attack, // <undefined> for cards other than minions
```
Comment (moollaza) [inline share/spice/hearthstone/hearthstone.js]: If there's a chance of these being undefined, I would first check to see if it's defined, and if so, in your normalize function, create the property and set the value accordingly

**Comment by Akryum** [comment]
@moollaza I don't really think of a good way of restricting the trigger... A keyword blacklist maybe?

**Comment by Akryum** [comment]
Using info template:
duckduckgo-hearthstone4

**Comment by Akryum** [comment]
I think that duckpan test modify my .travis.yml everytime. :/

File: lib/DDG/Spice/Hearthstone.pm [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+# Handle statement
+handle remainder => sub {
+
+    return unless $_;    # Guard against "no answer"
+    
+    # Keyword Blacklist
+    if(/(game|instruction|stoves|warcraft|deck|forum|wiki)/) {
+        return;
+    }
```
Comment (MrChrisW) [inline lib/DDG/Spice/Hearthstone.pm]: Can you place the guard keywords outside of the `handle` function like:

`my $keyword_guard = qr/game|instruction|stoves|warcraft|deck|forum|wiki/;`

Then inside:
`return if (/$keyword_guard/);`

Just for maintainability, easier to add new keywords.

**Comment by MrChrisW** [comment]
@moollaza Thanks for the comments. :+1: 
I did have a think about relevancy and my thought process was: 

- Check the query matches a list of cards (slurped in perl) - *unmaintainable (too many cards + adding new cards)*

- Modify the trigger to include explicitly the word "card" - *not user friendly*

A regex guard looks like the best way to go.

Just a note, the [mtg spice](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/Mtg.pm) doesn't include a regex guard and isn't currently live.

**Comment by Akryum** [comment]
Sometimes, the description of the card is cut too short:
duckduckgo-hearthstone5
It may have something to do with the info templates and handlebars.

**Comment by Akryum** [comment]
Resizing the window bring the full description back.
Also, the flavor text appears randomly for some cards and not for others.

**Comment by moollaza** [comment]
> I think that duckpan test modifies my .travis.yml all the time. :/

@Akryum Which DuckPAN test? You shouldn't need to commit a change to the .travis.yml

**Comment by Akryum** [comment]
Running the duckpan test command modifies the .travis.yml file automatically and deletes sudo:false each time.

**Comment by jagtalon** [comment]
@moollaza @Akryum yeah this happens because .travis.yml is autogenerated. Will make a PR for it!

**Comment by jagtalon** [comment]
Can be replicated with `dzil build` as well

**Comment by jagtalon** [comment]
@moollaza @Akryum https://github.com/duckduckgo/zeroclickinfo-spice/pull/1456

**Comment by jagtalon** [comment]
@Akryum we can't have bolded text in the IA (usually) since it's primary use is in pointing out the query in the search results.

**Comment by Akryum** [comment]
So do we keep it thus way?

**Comment by jagtalon** [comment]
@abeyang It's up on https://ddh1.duckduckgo.com/?q=hearthstone+leeroy+jenkins&ia=hearthstone

**Comment by jagtalon** [comment]
@Akryum maybe changing the color is better. @abeyang what do you think?

File: lib/DDG/Spice/Hearthstone.pm [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+handle remainder => sub {
+    
+    # Guard against "no answer"
+    return unless $_;
+    
+    # Keyword Blacklist
+    return if (/$keyword_guard/);
+    
+    # Regex guard
+    return $_ if (/^([a-z0-9':!\/,.-]+\s*)+$/i);
```
Comment (jagtalon) [inline lib/DDG/Spice/Hearthstone.pm]: What is it guarding against? It's a bit hard to tell from the regexp--could you add a comment that explains it. :)

File: lib/DDG/Spice/Hearthstone.pm [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+spice is_cached => 1;
+
+name "Hearthstone Card Search";
+source "Bytevortex (Gamepedia Proxy)";
+icon_url "http://eu.battle.net/hearthstone/static/images/icons/favicon.ico";
+description "Get a preview of any Hearthstone card.";
+primary_example_queries "hearthstone leeroy jenkins", "hearthstone leeroy", "leeroy hearthstone", "hearthstone mirror";
+category "entertainment";
+topics "gaming";
+code_url "https://github.com/Akryum/zeroclickinfo-spice/blob/master/lib/DDG/Spice/Hearthstone.pm";
```
Comment (jagtalon) [inline lib/DDG/Spice/Hearthstone.pm]: Please change to `https://github.com/duckduckgo/...`

File: share/spice/hearthstone/hearthstone.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                            value: api_result[key]
+                        });
+                    }
+                });
+                
+                // Normalized data
+                var card = {
+                    title: item.Has_name,
+                    url: "http://hearthstone.gamepedia.com/" + item.page,
+                    image: item.image_url,
```
Comment (jagtalon) [inline share/spice/hearthstone/hearthstone.js]: I wonder if we should maybe check the image_url? Is it always available?

**Comment by Akryum** [comment]
Imho, the color change may be a bit too subtle compared to bold.

**Comment by Akryum** [comment]
I added abbreviation for card keywords like "Battlecry" api-side, so that user can hover them and get a hint at what it does. No moficiation needed on the spice side.

**Comment by jkanarek** [comment]
@abeyang any ideas for a rollover style that looks a little less..."buttony"?

**Comment by moollaza** [comment]
@Akryum let's avoid any type of "rollover" effects. We don't typically use that for Instant Answers because it doesn't work on mobile devices.

**Comment by Akryum** [comment]
This is just a <abbr> HTML tag using the system tooltip, nothing special.

**Comment by jagtalon** [comment]
@Akryum Nice thanks. Yeah it might be too subtle--we'll get back to you on the style. In the mean time, here's a relevancy issue:

screen shot 2015-01-30 at 2 39 27 am

Maybe card should be stripped out in the `.pm` file.

**Comment by Akryum** [comment]
It should be fixed api-side (may need to empty cache).

**Comment by Akryum** [comment]
How is the spice cache emptied? Every day?

**Comment by moollaza** [comment]
@Akryum regarding the rollover, thanks for explaining, a tooltip should be okay, but if its important enough to show, I'd prefer we found a way that works for all devices.

The caching of results can be controlled by your IA based on the `spice proxy_cache_valid` setting. It defaults to:

`spice proxy_cache_valid => 200 1d`, which means we'll cache successful (200) API responses for one day. You can turn that higher or lower if you prefer, but we obviously lean towards more caching so we can return results faster.

**Comment by Akryum** [comment]
Well, for this kind of info we can go with a strong caching imho.

**Comment by Akryum** [comment]
@moollaza Is this correct? `spice proxy_cache_valid => "400 2d";`

**Comment by Akryum** [comment]
The cache was emptied so it's now okay on ddh1 : https://ddh1.duckduckgo.com/?q=hearthstone+leeroy+jenkins+card&ia=hearthstone

**Comment by abeyang** [comment]
Hi @Akryum, it's shaping up nicely. Is it possible to move the cost / attack / health to the card details (including the icons)?

And let's also unbold the description and un-italicize the footnote (and also use a smaller font and lighter color)...

pasted_image_2_2_15__12_46_pm

**Comment by Akryum** [comment]
Hi @abeyang, what's the correct way to add the icons in the infobox? (It escapes html)

**Comment by Akryum** [comment]
duckduckgo-hearthstone7

**Comment by Akryum** [comment]
This is the result for now:
duckduckgo-hearthstone7

**Comment by jagtalon** [comment]
@Akryum I actually like that. Looks super clean CC @abeyang

**Comment by abeyang** [comment]
@Akryum can you throw this onto DDH1? Would like to see this design with other cards (esp. those with a fuller description). But I think we're almost there!

**Comment by Akryum** [comment]
@abeyang Well, you may ask @jagtalon for that I think. :3

**Comment by jagtalon** [comment]
@Akryum @abeyang sorry about that. Updated: https://ddh1.duckduckgo.com/?q=hearthstone+leeroy+jenkins+card&ia=hearthstone

**Comment by jagtalon** [comment]
@abeyang I don't think we have `#999` as a standard color in https://duckduckgo.com/styleguide. What about using the `tx-clr--lt` class which adds `#888`?

screen shot 2015-02-06 at 3 49 01 am

**Comment by Akryum** [comment]
Flavor text nos has `tx-clr--lt` css class.

**Comment by moollaza** [comment]
@Akryum thanks!

@jagtalon can you update the code on DDH1 please? and then @abeyang can take another look and sign-off.

**Comment by jagtalon** [comment]
@abeyang @moollaza here you go! https://ddh9.duckduckgo.com/?q=hearthstone+tidecaller&ia=hearthstone looking good!

**Comment by moollaza** [comment]
Looks great! Would like a design sign-off and then we can merge it in :+1:

**Comment by abeyang** [comment]
Hi @Akryum , after looking at a few examples, I think it's best to just fit the little "gag" into the subtitle. For example, the [Crackle](https://ddh9.duckduckgo.com/?q=hearthstone+crackle&ia=hearthstone) card would look more like this:

> **Crackle**
> Snap! This card! Pop!
> 
> Deal 3-6 damage. Overload: (1)

Hope that makes sense. Let us know if you need help with anything!

**Comment by moollaza** [comment]
@abeyang unfortunately the Info template doesn't allow for a `subtitle` property.

We could add the markup to the description template, but unless we add additional CSS to shift it on the page it's going to look a little odd because the subtitle div will be contained in the `zci__content` div that wraps the description.

I think this has been fixed in the updated templates, so perhaps we can leave it as-is and change it later?

**Comment by chrismorast** [comment]
He means subheader @moollaza 
screen shot 2015-02-20 at 3 51 59 pm

**Comment by Akryum** [comment]
@chrismorast Do I put the flavor text subheader div in the content template or is there any property for that?

**Comment by chrismorast** [comment]
Hi @Akryum , put that content in the subtitle property. We will be deploying an update to the template to allow for this within the next day or two.

**Comment by moollaza** [comment]
@Akryum sorry! I didn't realize you had updated this -- we don't get notified of commits, only comments!

**Comment by moollaza** [comment]
LGTM :+1:

**Comment by moollaza** [comment]
@Akryum I just noticed we're not using the CSS file -- can you make a PR to remove it?

**Comment by moollaza** [comment]
Nevermind I see what's happening. The text comes as HTML from the API. Probably best to use `DDG.stripHTML()` but this is fine.

**Comment by moollaza** [comment]
@Akryum congrats! It's live: https://duckduckgo.com/?q=hearthstone+murloc+warleader&ia=hearthstone

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Comment by Akryum** [comment]
Hi! It seems that the card IA is no longer working. Is there any problem I can help with?

**Comment by moollaza** [comment]
@Akryum do you have any examples? I just tried and it worked:

hearthstone_murloc_warleader_at_duckduckgo_and_duckduckgo_-_hipchat

**Comment by Akryum** [comment]
@moollaza It seems that some cards no longer display the IA like these:
https://duckduckgo.com/?q=hearthstone+blizzard
https://duckduckgo.com/?q=hearthstone+oasis
https://duckduckgo.com/?q=hearthstone+boom+bot

I just tested my api endpoint and it got quite slow now. Has the IA a timeout? (I will see what I can do to speed up the api.)

**Comment by Akryum** [comment]
I have added a cache system to greatly speed up calls to my api so the IA works again after a few try (when the card cache is ready).

**Comment by moollaza** [comment]
@Akryum yes there's a default timeout of 5000ms. 

Your 3 examples all worked for me, so I guess everything is good now?

**Comment by Akryum** [comment]
Yes, evrything is working again :)

**Pull request merged**